### PR TITLE
implement SignUP

### PR DESCRIPTION
**Pull request by mrcuong84**

### CODE REVIEW

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -76,6 +76,25 @@ footer {
...
+  width: 100%;
+  margin-bottom: 20px;
+
+}
+
+input {
+  height: auto !important;
+}
+
+.avatar{
```
Comment (dinhvareal) [inline app/assets/stylesheets/custom.scss]: .custom-avatar

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -76,6 +76,25 @@ footer {
...
+  width:90px;
```
Comment (dinhvareal) [inline app/assets/stylesheets/custom.scss]: width: 150px;

File: app/views/users/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<div class="row">
+  <div class="col-md-3 col-md-offset-3">
+    <h1>Sign up</h1>
+    <%= form_for(@user) do |f| %>
+
+      <%= render 'kaminari/error_messages' %>
```
Comment (dinhvareal) [inline app/views/users/new.html.erb]: move error_message to shared

Comment (mrcuong84) [reply app/views/users/new.html.erb]: resolved

File: app/assets/stylesheets/custom.scss (hunk 1 above)
Comment (mrcuong84) [reply app/assets/stylesheets/custom.scss]: already fix

File: app/assets/stylesheets/custom.scss (hunk 2 above)
Comment (mrcuong84) [reply app/assets/stylesheets/custom.scss]: resolved

File: app/views/users/new.html.erb (hunk 3 above)
Comment (mrcuong84) [reply app/views/users/new.html.erb]: resolved

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by(phone: params[:session][:phone])
+    if user &.authenticate(params[:session][:password])
+      log_in user
+      flash[:success] = 'Dang nhap thanh cong!'
```
Comment (hieunguyentrung) [inline app/controllers/sessions_controller.rb]: I18n ?

File: Gemfile [hunk 5]
```diff
@@ -15,6 +15,12 @@ gem 'rails', '~> 5.2.3'
 gem 'bootstrap-sass', '~> 3.4.1'
 gem 'sassc-rails', '>= 2.1.0'
 
+gem 'bcrypt',         '3.1.12'
+
+gem 'carrierwave',             '1.2.2'
```
Comment (hieunguyentrung) [inline Gemfile]: Please remove.

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (mrcuong84) [reply app/controllers/sessions_controller.rb]: resolved

File: Gemfile (hunk 5 above)
Comment (mrcuong84) [reply Gemfile]: already removed

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+class UsersController < ApplicationController
+  def index
+    @user = User.all
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
+    return if @user.present?
+
+    flash[:danger] = 'User khong ton tai, moi ban dang nhap'
```
Comment (hieunguyentrung) [inline app/controllers/users_controller.rb]: I18n ?

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -9,6 +9,10 @@
   </head>
   <body>
     <%= render 'layouts/header' %>
+    <div class="container custom-body-html">
+      <% flash.each do |message_type, message| %>
```
Comment (hieunguyentrung) [inline app/views/layouts/application.html.erb]: Duplicate

File: app/views/sessions/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      <%= f.phone_field :phone, class: 'form-control' %>
+    </div>
+
+    <div class="form-group">
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+    </div>
+
+    <div class="form-check">
+      <%= f.submit "Log in", class: "btn btn-primary" %>
```
Comment (hieunguyentrung) [inline app/views/sessions/new.html.erb]: I18n.

File: app/views/sessions/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    </div>
+
+    <% end %>
+
+    <p>New user? <%= link_to "Sign up now!", signup_path %></p>
```
Comment (hieunguyentrung) [inline app/views/sessions/new.html.erb]: I18n.

File: app/views/shared/_error_messages.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, "error") %>.
```
Comment (hieunguyentrung) [inline app/views/shared/_error_messages.html.erb]: I18n.

File: app/views/shared/_error_messages.html.erb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+    </div>
+    <ul>
+    <% @user.errors.full_messages.each do |msg| %>
```
Comment (hieunguyentrung) [inline app/views/shared/_error_messages.html.erb]: Change `msg` to `message`.

File: app/views/users/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    <%= form_for(@user, html: {class: 'form-group'} ) do |f| %>
+
+      <%= render 'shared/error_messages' %>
+      <div class="form-group">
+        <%= f.label :name %>
+        <%= f.text_field :name, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :phone %> <p><i>(10 chữ số định dạng 09xxx.....)</i></p>
```
Comment (hieunguyentrung) [inline app/views/users/new.html.erb]: I18n.

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        <%= f.phone_field :phone, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :address %>
+        <%= f.text_field :address, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :password %> <p><i>(nhập ít nhất 6 ký tự)</i></p>
```
Comment (hieunguyentrung) [inline app/views/users/new.html.erb]: I18n.

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        <%= f.password_field :password, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :password_confirmation, "Password Confirmation" %>
+        <%= f.password_field :password_confirmation, class: 'form-control'%>
+      </div>
+
+      <div class="form-check">
+        <%= f.submit "Create my account", class: "btn btn-primary" %>
```
Comment (hieunguyentrung) [inline app/views/users/new.html.erb]: I18n.

File: app/views/users/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        <%= f.text_field :address, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :password %> <p><i>(nhập ít nhất 6 ký tự)</i></p>
+        <%= f.password_field :password, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :password_confirmation, "Password Confirmation" %>
```
Comment (hieunguyentrung) [inline app/views/users/new.html.erb]: I18n.

File: db/migrate/20190711013707_create_users.rb [hunk 16]
```diff
@@ -0,0 +1,13 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :password_digest
+      t.string :phone
```
Comment (hieunguyentrung) [inline db/migrate/20190711013707_create_users.rb]: `NOT NULL`

File: config/routes.rb [hunk 17]
```diff
@@ -1,4 +1,12 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
+  get 'users/new'
+  get '/signup', to: 'users#new'
```
Comment (hieunguyentrung) [inline config/routes.rb]: Just using `/users/new` from `users` resource. Don't need `/signup` path.

File: config/routes.rb [hunk 18]
```diff
@@ -1,4 +1,12 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (hieunguyentrung) [inline config/routes.rb]: This code is redundant.

File: config/routes.rb [hunk 19]
```diff
@@ -1,4 +1,12 @@
...
+  get 'users/new'
```
Comment (hieunguyentrung) [inline config/routes.rb]: This code is redundant.

File: config/routes.rb [hunk 20]
```diff
@@ -1,4 +1,12 @@
...
+  get '/signup', to: 'users#new'
+  get '/login', to: 'sessions#new'
+  post '/login', to: 'sessions#create'
```
Comment (hieunguyentrung) [inline config/routes.rb]: Using default `POST /users`.

File: config/routes.rb [hunk 21]
```diff
@@ -1,4 +1,12 @@
...
+  delete '/logout',  to: 'sessions#destroy'
```
Comment (hieunguyentrung) [inline config/routes.rb]: Using default path when using `resource :users`.

File: config/routes.rb [hunk 22]
```diff
@@ -1,4 +1,12 @@
...
+  post '/signup',  to: 'users#create'
```
Comment (hieunguyentrung) [inline config/routes.rb]: Using default path when using `resource :users`.

File: db/migrate/20190711013707_create_users.rb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :password_digest
```
Comment (hieunguyentrung) [inline db/migrate/20190711013707_create_users.rb]: `NOT NULL`

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: app/controllers/sessions_controller.rb [hunk 24]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by(phone: params[:session][:phone])
+    if user&.authenticate(params[:session][:password])
+      log_in user
+      flash[:success] = 'Đăng nhập thành công'
```
Comment (dinhvareal) [inline app/controllers/sessions_controller.rb]: I18n

File: app/controllers/sessions_controller.rb [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+      redirect_to user
+    else
+      flash[:warning] = 'Nhập sai Số điện thoại/password'
```
Comment (dinhvareal) [inline app/controllers/sessions_controller.rb]: I18n

File: app/views/sessions/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,24 @@
+<div class="row">
+  <h1>Log in</h1>
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path, html: {class: 'form-group'} ) do |f| %>
+
+    <div class="form-group">
+      <%= f.label :phone %>
```
Comment (dinhvareal) [inline app/views/sessions/new.html.erb]: you can change :phone to t('.phone') for the label

File: app/views/sessions/new.html.erb [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+      <%= f.phone_field :phone, class: 'form-control' %>
+    </div>
+
+    <div class="form-group">
+      <%= f.label :password %>
```
Comment (dinhvareal) [inline app/views/sessions/new.html.erb]: Use I18n

File: app/views/users/new.html.erb [hunk 28]
```diff
@@ -0,0 +1,38 @@
+<div class="row">
+  <div class="col-md-4 col-md-offset-3">
+    <h1>Sign up</h1>
+    <%= form_for(@user, html: {class: 'form-group'} ) do |f| %>
+
+      <%= render 'shared/error_messages' %>
+      <div class="form-group">
+        <%= f.label :name %>
```
Comment (dinhvareal) [inline app/views/users/new.html.erb]: Use I18n for the label

File: app/views/users/new.html.erb [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+        <%= f.text_field :name, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :phone %> <p><i>(10 chữ số định dạng 09xxx.....)</i></p>
+        <%= f.phone_field :phone, class: 'form-control' %>
+      </div>
+
+      <div class="form-group">
+        <%= f.label :address %>
```
Comment (dinhvareal) [inline app/views/users/new.html.erb]: Use I18n for the label

File: app/views/users/index.html.erb [hunk 30]
```diff
@@ -0,0 +1,12 @@
+<% @user.each do |user| %>
+  <div class="col-md-6">
+    <div class="col-md-12 custom-book-col">
+      <div class="col-sm-3"><img src="https://media.istockphoto.com/vectors/bookstore-logo-template-design-logo-open-book-vector-id1088265598" alt="Img" class="avatar"></div>
```
Comment (dinhvareal) [inline app/views/users/index.html.erb:4]: img src="<%= user.avatar %>"

**Review by hieunguyentrung — APPROVED** [review]

**Pull request merged**